managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java

```diff
@@ -2793,30 +2794,23 @@ public void skipNonRecoverableLedger(final long ledgerId){
         if (ledgerInfo == null) {
             return;
         }
-        lock.writeLock().lock();
         log.warn("[{}] [{}] Since the ledger [{}] is lost and the autoSkipNonRecoverableData is true, this ledger will"
                 + " be auto acknowledge in subscription", ledger.getName(), name, ledgerId);
-        try {
-            for (int i = 0; i < ledgerInfo.getEntries(); i++) {
-                if (!individualDeletedMessages.contains(ledgerId, i)) {
-                    asyncDelete(PositionImpl.get(ledgerId, i), new AsyncCallbacks.DeleteCallback() {
-                        @Override
-                        public void deleteComplete(Object ctx) {
-                            // ignore.
-                        }
+        asyncDelete(() -> LongStream.range(0, ledgerInfo.getEntries())
```

Comment:
Improve the code like below is better

```java
lock.readLick().lock();
List posToDelete = new();
for (int i = 0; i < ledgerInfo.getEntries(); i++) {
    if (!individualDeletedMessages.contains(ledgerId, i)) {
        posToDelete.add();
    }
}
lock.readLick().unlock();

asyncDelete(posToDelete);
```

Reply:
I think the current approach is better because the method you mentioned involves an additional locking operation, and it creates all the Position objects in advance before executing asyncDelete, which would lead to more memory consumption, right?

Reply:
@hrzzzz 

> It creates all the Position objects before executing asyncDelete, which would lead to more memory consumption, right?

Regarding the current implementation, it also creates Position instances, right?

For example: 
- entries count is `1,000`
- asked `500`
- the difference between both solutions
  - current solution: will create `1000` position instances
  - the solution I suggested will create `500` position instances, and it is faster

Reply:
Creating a PositionImpl object is a very light operation. Especially in the streaming approach where they live for a very short time, single loop instead of being created beforehand. 

Adding a deletion check is the position has been created is most likely more expensive than creating the object. The deletion check is any ways done in the asyncDelete method and it's extra work to have the pre-check.

Reply:
> Adding a deletion check is the position has been created is most likely more expensive than creating the object. The deletion check is any ways done in the asyncDelete method and it's extra work to have the pre-check.

Sure